contracts/QuestNFT.sol

```diff
@@ -0,0 +1,219 @@
+    function totalTransferAmount() external view returns (uint256) {
+        return questFee * totalParticipants;
+    }
+
+    /// @notice Pauses the Quest
+    /// @dev Only the owner of the Quest can call this function.
+    function pause() external onlyOwner {
+        _pause();
+    }
+
+    /// @notice Unpauses the Quest
+    /// @dev Only the owner of the Quest can call this function.
+    function unPause() external onlyOwner {
+        _unpause();
+    }
+
+    function safeMint(address to_) public onlyMinter onlyQuestBetweenStartEnd whenNotPaused nonReentrant {
+        _tokenIdCounter.increment();
+        uint tokenId = _tokenIdCounter.current();
+        _safeMint(to_, tokenId);
+        (bool success, ) = protocolFeeRecipient.call{value: questFee}("");
+        require(success, 'protocol fee transfer failed');
+    }
+
+    function _beforeTokenTransfer(address from, address to, uint256 tokenId, uint256 batchSize)
+        internal
+        whenNotPaused
+        override
+    {
+        super._beforeTokenTransfer(from, to, tokenId, batchSize);
+    }
+
+    /// @dev Function that to withdraw the remaining coins in the contract to the owner
+    /// @notice This function can only be called after the quest end time
+    function withdrawRemainingTokens() external onlyAfterQuestEnd {
+        uint balance = address(this).balance;
+        if (balance > 0) {
+            (bool success, ) = owner().call{value: balance}("");
+            require(success, 'withdraw remaining tokens failed');
+        }
+
+    }
+
+    /// @dev saftey hatch function to transfer tokens sent to the contract to the contract owner.
+    /// @param erc20Address_ The address of the ERC20 token to refund
+    function refund(address erc20Address_) external {
+        uint erc20Balance = IERC20(erc20Address_).balanceOf(address(this));
+        if (erc20Balance > 0) IERC20(erc20Address_).safeTransfer(owner(), erc20Balance);
+    }
+
+    /// @dev returns the token uri
+    /// @param tokenId_ the token id
+    function tokenURI(uint256 tokenId_)
+        public
+        view
+        override(ERC721Upgradeable)
+        returns (string memory)
+    {
+        bytes memory dataURI = generateDataURI();
+        return string(abi.encodePacked('data:application/json;base64,', Base64.encode(dataURI)));
+    }
+
+    /// @dev returns the data uri in json format
+    function generateDataURI() internal view virtual returns (bytes memory) {
+        bytes memory dataURI = abi.encodePacked(
+            '{',
+            '"name": "',
+            name(),
+            '",',
+            '"description": "',
+            description,
+            '",',
+            '"image": "',
+            tokenImage(),
+            '"',
+            '}'
+        );
+        return dataURI;
+    }
+
+    function tokenImage() internal view virtual returns (string memory) {
+        return string(abi.encodePacked('ipfs://', imageIPFSHash));
```

Comment:
Should we have a `setTokenImage` function? Or is intention for it to be permanent?